### PR TITLE
Create /locate and /compass commands

### PR DESCRIPTION
**Pull request by magicus**
Usage:

`/locate place|service <name>` 
shows a list of matches. Clicking on a match sets the compass. `<name>` is matched using the partial match so "dtls" matches "Detlas" and "Detlas Suburbs", while "dtlsu" only matches "Detlas Suburbs"

`/compass place|service <name>`
Similar to locate, but automatically sets the compass to the matching place, if a unique place could be determined (either by fuzzy matching only a single hit, or by having an exact match, so "Detlas" is ok and will not be considered a partial match of "Detlas Suburbs"). For services, the closest service is selected.

`/compass at <location>`
Uses vanilla coordinates and sets the compass to that exact coordinates, e.g. `/compass at ~ ~ ~` sets the compass at your current location.

### CODE REVIEW

File: common/src/main/java/com/wynntils/commands/CompassCommand.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                        .build())
+                .executes(this::syntaxError);
+    }
+
+    private int compassAt(CommandContext<CommandSourceStack> context) {
+        Coordinates coordinates = Vec3Argument.getCoordinates(context, "location");
+        Location location = new Location(coordinates.getBlockPos(context.getSource()));
+        CompassModel.setCompassLocation(location);
+
+        MutableComponent response = new TextComponent("Setting compass to ").withStyle(ChatFormatting.AQUA);
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/commands/CompassCommand.java]: maybe "Set compass to x y z" is better, as it isnt really a continuous process

File: common/src/main/java/com/wynntils/commands/CompassCommand.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            // This really should not happen...
+            MutableComponent response = new TextComponent("Found no services of kind '" + selectedKind.getName() + "'")
+                    .withStyle(ChatFormatting.RED);
+            context.getSource().sendFailure(response);
+            return 0;
+        }
+        Poi closestService = closestServiceOptional.get();
+        CompassModel.setCompassLocation(closestService.getLocation().asLocation());
+
+        MutableComponent response = new TextComponent("Setting compass to " + selectedKind.getName() + " at ")
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/commands/CompassCommand.java]: Same comment as above

File: common/src/main/java/com/wynntils/commands/CompassCommand.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                .withStyle(ChatFormatting.AQUA);
+        response.append(new TextComponent(closestService.getLocation().toString()).withStyle(ChatFormatting.WHITE));
+        context.getSource().sendSuccess(response, false);
+        return 1;
+    }
+
+    private int compassPlace(CommandContext<CommandSourceStack> context) {
```
Comment (Incompleteusern) [inline common/src/main/java/com/wynntils/commands/CompassCommand.java:124]: This is really similar to `/territory`

File: common/src/main/java/com/wynntils/commands/CompassCommand.java (hunk 1 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/commands/CompassCommand.java]: That sounds like an imperative to the user... Maybe "Compass set to ..."?

Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/commands/CompassCommand.java]: Yeah

File: common/src/main/java/com/wynntils/commands/CompassCommand.java (hunk 3 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/commands/CompassCommand.java:124]: I'm planning to look how I can unite the territory code with the new MapModel. The difference here is the selection of POIs that this draws from; this here is place names from the Wynncraft web map, not territories.

File: common/src/main/java/com/wynntils/commands/LocateCommand.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                        && servicePoi.getKind().equals(selectedKind))
+                .toList());
+
+        // Only keep the closest results
+        Vec3 currentLocation = McUtils.player().position();
+        services.sort(Comparator.comparingDouble(poi -> currentLocation.distanceToSqr(
+                poi.getLocation().getX(),
+                poi.getLocation().getY(),
+                poi.getLocation().getZ())));
+        services.subList(4, services.size()).clear();
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/commands/LocateCommand.java:105]: What is this?

File: common/src/main/java/com/wynntils/wynn/model/map/poi/MapLocation.java [hunk 5]
```diff
@@ -26,4 +28,17 @@ public int getY() {
...
+        return new Location(x, y, z);
+    }
+
+    @Override
+    public String toString() {
+        return "[" + x + ", " + y + ", " + z + "]";
+    }
+
+    public String asChatCoordinates() {
+        return x + " " + y + " " + z;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/wynn/model/map/poi/MapLocation.java:42]:
```suggestion
        return x + ", " + y + ", " + z;
```

Perhaps this is better?

**Review by kristofbolyai — DISMISSED** [review]
Seems good overall

File: common/src/main/java/com/wynntils/commands/LocateCommand.java (hunk 4 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/commands/LocateCommand.java:105]: Remove everything except the three first elements in the list. As suggested by Stack Overflow.

File: common/src/main/java/com/wynntils/wynn/model/map/poi/MapLocation.java (hunk 5 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/wynn/model/map/poi/MapLocation.java:42]: No, this is "chat coordinates", that is, how Vanilla expects a coordinate to be written. This way I can use vanilla functionality. 

When normally presenting to user, I have the toString method with commas.

I also intend to make a more forgiving input method, like in Legacy, for anything "coordinate like". I'm not just there yet.

File: common/src/main/java/com/wynntils/commands/LocateCommand.java (hunk 4 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/commands/LocateCommand.java:105]: Ah.., makes sense. Maybe I would have figured out after staring for like 10 minutes. I just spent 1 hour wondering why my POI rendering trigonometry did not work... (I centered my math around the screen and not the map center.... :facepalm: )

**Review by kristofbolyai — APPROVED** [review]
Just as a note, compass share is a todo. I know you won't forget that though.

**Pull request merged**